#MultiColorLayout does formatting, parsing, coloring and managing human-readable color names and should be refactored.

piddubnyi opened the pull request:
Solving #59 
Extracted Color Manager so now managing human-readable color names is done in a separate class.

dmarkov commented:
@piddubnyi Thanks for the pull request, let me find a reviewer..

dmarkov commented:
@mkordas could you please review this pull request

mkordas commented:
@piddubnyi I'm on it

mkordas commented on src/main/java/com/jcabi/log/ColorManager.java:
```diff
@@ -0,0 +1,97 @@
+ * Color manager to store color data.
+ * @author Igor Piddubnyi (igor.piddubnyi@gmail.com)
+ * @version $Id$
+ * @since 1.0
```
@piddubnyi should be `@since 0.18`

mkordas commented on src/main/java/com/jcabi/log/ColorManager.java:
```diff
@@ -0,0 +1,97 @@
+ * @version $Id$
+ * @since 1.0
+ */
+public class ColorManager {
```
@piddubnyi please read http://www.yegor256.com/2015/03/09/objects-end-with-er.html and watch https://www.youtube.com/watch?v=WpP4rIhh5e4

mkordas commented on src/main/java/com/jcabi/log/ColorManager.java:
```diff
@@ -0,0 +1,97 @@
+
+    /**
+     * Add color to color map.
+     * @param key Key to add.
```
@piddubnyi let's be consistent and not add dots at the end of at-tags as you did a few lines below

mkordas commented on src/main/java/com/jcabi/log/ColorManager.java:
```diff
@@ -0,0 +1,97 @@
+     */
+    private static ConcurrentMap<String, String> colorMap() {
+        final ConcurrentMap<String, String> map =
+            new ConcurrentHashMap<String, String>();
```
@piddubnyi We're not modifying that map after creation, I think we should use `ImmutableMap` instead

mkordas commented on src/main/java/com/jcabi/log/ColorManager.java:
```diff
@@ -0,0 +1,97 @@
+     * Colors with names.
+     */
+    private final transient ConcurrentMap<String, String> colors =
+        ColorManager.colorMap();
```
@piddubnyi please move that field initialization to constructor

mkordas commented on src/main/java/com/jcabi/log/MulticolorLayout.java:
```diff
@@ -75,7 +75,7 @@
  *  &lt;artifactId&gt;jcabi-log&lt;/artifactId&gt;
  * &lt;/dependency&gt;</pre>
- *
+ * @todo #59:30min Continue refactoring of the class started in #51.
```
@piddubnyi can you provide more details here? Like what was already done and what is still left?

mkordas commented on src/main/java/com/jcabi/log/MulticolorLayout.java:
```diff
@@ -137,6 +131,11 @@
     private transient String base;
 
+    /**
+     * Helper class to store color data.
```
@piddubnyi
> Helper class

Please look at http://blogs.msdn.com/b/nickmalik/archive/2005/09/06/461404.aspx and http://www.smart421.com/application-development/why-helper-singletons-and-utility-classes-are-mostly-bad-2/

mkordas replied:
@piddubnyi please replace 'helper class' with something different

mkordas commented:
@piddubnyi please see a few comments above

piddubnyi commented:
@mkordas I dont see an easy way to decouple logic of this class, that's why I did first and easiest thing: I moved logic for managing human readable color names to another class, and Introduced a puzzle to continue on refactoring.
What is your opinion, how can we solve initial issue #59 ?

mkordas commented:
@piddubnyi that approach of extracting code about colors is fine, but again, please read http://www.yegor256.com/2015/03/09/objects-end-with-er.html and watch https://www.youtube.com/watch?v=WpP4rIhh5e4

mkordas commented:
@piddubnyi moreover, please try to fix the build

mkordas commented on src/main/java/com/jcabi/log/ColorManager.java:
```diff
@@ -0,0 +1,103 @@
+    /**
+     * Colors with names.
+     */
+    private transient ConcurrentMap<String, String> colors;
```
@piddubnyi shuld be `final`

mkordas commented:
@piddubnyi introducing a puzzle is good approach to partially solve issue for now, but you need to put more description there

mkordas commented:
@piddubnyi actually build is failing because of the following:
```
[ERROR] \src\main\java\com\jcabi\log\ColorManager.java[46-46]: Private field 'colors' could be made final; it is only initialized in the declaration or constructor. (ImmutableField)
```

piddubnyi commented:
@mkordas Please check.

mkordas commented on src/main/java/com/jcabi/log/Colors.java:
```diff
@@ -0,0 +1,103 @@
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Color manager to store color data.
```
@piddubnyi please adjust Javadoc

mkordas commented on src/main/java/com/jcabi/log/MulticolorLayout.java:
```diff
@@ -75,7 +75,8 @@
  *  &lt;artifactId&gt;jcabi-log&lt;/artifactId&gt;
  * &lt;/dependency&gt;</pre>
- *
+ * @todo #59:30min This class is still handling multiple responsibilities like
```
@piddubnyi `@todo` tag should be placed at the end of Javadoc

mkordas commented on src/main/java/com/jcabi/log/MulticolorLayout.java:
```diff
@@ -137,6 +132,11 @@
+    /**
+     * Helper class to store color data.
+     */
+    private final transient Colors colrs = new Colors();
```
@piddubnyi typo: should be `colors`

mkordas commented:
@piddubnyi please see a few comments above

piddubnyi commented:
@caarlos0 Fixed.

caarlos0 commented:
@piddubnyi ?

piddubnyi commented:
@caarlos0 sorry, wrong name. I meant @mkordas

mkordas commented:
@piddubnyi good job!

mkordas commented:
@rultor merge

rultor commented:
> @rultor merge

@mkordas Thanks for your request. @yegor256 Please confirm this.

yegor256 commented:
@rultor try to merge

rultor commented:
> @rultor try to merge

@yegor256 OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/6648-176954507)

rultor commented:
> @rultor try to merge

@yegor256 Done! FYI, the full log is [here](http://www.rultor.com/t/6648-176954507) (took me 6min)

mkordas commented:
@rultor deploy

rultor commented:
> @rultor deploy

@mkordas Thanks for your request. @yegor256 Please confirm this.

mkordas commented:
@yegor256 ping, it's waiting for merge 4 days already

yegor256 commented:
@rultor merge

rultor commented:
> @rultor merge

@piddubnyi @yegor256 The pull request is closed already, so I can't merge it

rultor commented:
> @rultor merge

@yegor256 I'm sorry, I don't understand you :( Check [this page](http://doc.rultor.com/basics.html) and try again please

yegor256 commented:
@rultor deploy

rultor commented:
> @rultor deploy

@yegor256 OK, I'll try to deploy now. You can check the progress [here](http://www.rultor.com/t/6648-178906084)

rultor commented:
> @rultor deploy

@piddubnyi @yegor256 Oops, I failed. You can see the full log [here](http://www.rultor.com/t/6648-178906084) (spent 3min)

```
eeb870fa77f6: Already exists
8e91dd8e64a9: Already exists
7edc9591cfff: Already exists
b6f0ef70cdda: Already exists
55f50dadf6df: Already exists
5bb70c50a08e: Already exists
3e96c48e0103: Already exists
af499a0b9fef: Already exists
8347d5c27a07: Already exists
91032581ae21: Already exists
e14f7ea174ff: Already exists
42f4b22c0643: Already exists
0f3bc2151c35: Already exists
9c9ef197ca57: Already exists
e915b8214106: Already exists
b1aa9b40b130: Already exists
3ef9ddc66ccf: Already exists
aaec00c5eddd: Already exists
3973129a19af: Already exists
d2118f9af91b: Already exists
58438bb4378b: Already exists
889de6b692b8: Already exists
123213ef0aca: Already exists
861c99bede6c: Already exists
fea2902e32c2: Already exists
fe63f5e653d4: Already exists
a3993bd96606: Already exists
afa03f817a39: Already exists
1eefa2535ae7: Already exists
490717c804fe: Already exists
Digest: sha256:b7b81dbc5c2e8bb74d336d568b84a6f4b903bca608c980c6ae231b8a46f5e922
Status: Image is up to date for docker.io/yegor256/rultor:latest

+ docker ps --filter=status=exited
+ grep --quiet '\sjcabi_jcabi-log_72\s*$'
++ pwd
++ pwd
+ docker run --rm -v /tmp/rultor-c9gB:/main '--env=MAVEN_OPTS=-XX:MaxPermSize=256m -Xmx1g' --env=head=git@github.com:jcabi/jcabi-log.git --env=head_branch=master --privileged=true --hostname=docker --memory=6g --memory-swap=16g --oom-kill-disable --cidfile=/tmp/rultor-c9gB/cid -w=/main --name=jcabi_jcabi-log_72 yegor256/rultor /main/entry.sh
+ set -e
+ set -o pipefail
+ shopt -s dotglob
+ useradd -m -G sudo r
+ echo '%sudo ALL=(ALL) NOPASSWD:ALL'
+ cp -R /root/.bashrc /root/.composer /root/.gem /root/.java /root/.m2 /root/.npm /root/.profile /root/texmf /root/.texmf-var /root/tmp /home/r
+ cp -R ./cid ./end.sh ./entry.sh ./pid ./pubring.gpg ./repo ./run.sh ./s3cfg ./script.sh ./secring.gpg ./settings.xml ./stdout /home/r
+ rm -rf repo
+ chown -R r:r /home/r
+ chmod a+x /home/r/script.sh
+ su -m r -c /home/r/script.sh
+ set -e
+ set -o pipefail
+ export HOME=/home/r
+ HOME=/home/r
+ cd /home/r/repo
+ sudo gem install pdd
ERROR:  Could not find a valid gem 'pdd' (>= 0), here is why:
          Unable to download data from https://rubygems.org/ - no such name (https://rubygems.org/specs.4.8.gz)
container c5badf55a2a918417a07f5a536753d76b966766e27a7de519af1d32c81015627 is dead
Wed Feb  3 00:27:51 UTC 2016

```

yegor256 commented:
@rultor try to merge

rultor commented:
> @rultor try to merge

@piddubnyi @yegor256 The pull request is closed already, so I can't merge it

rultor commented:
> @rultor try to merge

@yegor256 I'm sorry, I don't understand you :( Check [this page](http://doc.rultor.com/basics.html) and try again please

mkordas commented:
@rultor deploy

rultor commented:
> @rultor deploy

@mkordas Thanks for your request. @yegor256 Please confirm this.

yegor256 commented:
@rultor deploy

rultor commented:
> @rultor deploy

@yegor256 OK, I'll try to deploy now. You can check the progress [here](http://www.rultor.com/t/6648-179432352)

rultor commented:
> @rultor deploy

@yegor256 Done! FYI, the full log is [here](http://www.rultor.com/t/6648-179432352) (took me 5min)

dmarkov commented:
@mkordas paid, thanks, added **38 mins** to your account, payment `56b31b084f694f4168000195`, 119 hours and 18 mins was spent; extra minutes for review comments (c=23); +38 added to your rating, current score is: [+5002](http://www.netbout.com/b/36080?open=rating)

dmarkov commented:
@rultor deploy

rultor commented:
> @rultor deploy

@dmarkov OK, I'll try to deploy now. You can check the progress [here](http://www.rultor.com/t/6648-179731807)

rultor commented:
> @rultor deploy

@dmarkov Done! FYI, the full log is [here](http://www.rultor.com/t/6648-179731807) (took me 6min)
